File: week1/factory.java
public class factory {
    interface MyDoc {
        void open();
    }

    static class WordDoc implements MyDoc {
        public void open() {
            System.out.println("Word file is now open");
        }
    }

    static class PdfDoc implements MyDoc {
        public void open() {
            System.out.println("PDF file is now open");
        }
    }

    static class ExcelDoc implements MyDoc {
        public void open() {
            System.out.println("Spreadsheet opened.");
        }
    }

    abstract static class DocMaker {
        abstract MyDoc makeDoc();
    }

    static class WordDocMaker extends DocMaker {
        public MyDoc makeDoc() {
            return new WordDoc();
        }
    }

    static class PdfDocMaker extends DocMaker {
        public MyDoc makeDoc() {
            return new PdfDoc();
        }
    }

    static class ExcelDocMaker extends DocMaker {
        public MyDoc makeDoc() {
            return new ExcelDoc();
        }
    }

    public static void main(String[] args) {
        DocMaker doc = new WordDocMaker();
        MyDoc word = doc.makeDoc();
        word.open();

        DocMaker pdf = new PdfDocMaker();
        MyDoc p = pdf.makeDoc();
        p.open();

        DocMaker excel = new ExcelDocMaker();
        MyDoc x = excel.makeDoc();
        x.open();
    }
}

File: week1/Factory_method/factory.cs
using System;

public class Factory
{
    public interface IMyDoc
    {
        void Open();
    }

    public class WordDoc : IMyDoc
    {
        public void Open()
        {
            Console.WriteLine("Word file is now open");
        }
    }

    public class PdfDoc : IMyDoc
    {
        public void Open()
        {
            Console.WriteLine("PDF file is now open");
        }
    }

    public class ExcelDoc : IMyDoc
    {
        public void Open()
        {
            Console.WriteLine("Spreadsheet opened.");
        }
    }

    public abstract class DocMaker
    {
        public abstract IMyDoc MakeDoc();
    }

    public class WordDocMaker : DocMaker
    {
        public override IMyDoc MakeDoc()
        {
            return new WordDoc();
        }
    }

    public class PdfDocMaker : DocMaker
    {
        public override IMyDoc MakeDoc()
        {
            return new PdfDoc();
        }
    }

    public class ExcelDocMaker : DocMaker
    {
        public override IMyDoc MakeDoc()
        {
            return new ExcelDoc();
        }
    }

    public static void Main()
    {
        DocMaker wordMaker = new WordDocMaker();
        IMyDoc word = wordMaker.MakeDoc();
        word.Open();

        DocMaker pdfMaker = new PdfDocMaker();
        IMyDoc pdf = pdfMaker.MakeDoc();
        pdf.Open();

        DocMaker excelMaker = new ExcelDocMaker();
        IMyDoc excel = excelMaker.MakeDoc();
        excel.Open();
    }
}
